testsuite/tests/typing-modular-explicits/general.ml

```diff
@@ -0,0 +1,1031 @@
+(* TEST
+  expect;
+*)
+
+module type Typ = sig type t end
+
+module type Add = sig type t val add : t -> t -> t end
+
+let id (module T : Typ) (x : T.t) = x
+
+let id2 : (module T : Typ) -> T.t -> T.t =
+  fun (module A : Typ) (x : A.t) -> x
+
+let id_infer_sig : (module T : Typ) -> T.t -> T.t =
+  fun (module A) (x : A.t) -> x
+
+[%%expect{|
+module type Typ = sig type t end
+module type Add = sig type t val add : t -> t -> t end
+val id : (module T : Typ) -> T.t -> T.t = <fun>
+val id2 : (module T : Typ) -> T.t -> T.t = <fun>
+val id_infer_sig : (module T : Typ) -> T.t -> T.t = <fun>
+|}]
+
+
+let f x y = (id (module Int) x, id (module Bool) y)
+
+[%%expect{|
+val f : Int.t -> Bool.t -> Int.t * Bool.t = <fun>
+|}]
+
+let f2 x y = (id (module Int : Typ) x, id (module Bool : Typ) y)
+
+[%%expect{|
+val f2 : Int.t -> Bool.t -> Int.t * Bool.t = <fun>
+|}]
+
+let merge (module T : Typ) x y = (id (module T) x, id (module T) y)
+
+[%%expect{|
+val merge : (module T : Typ) -> T.t -> T.t -> T.t * T.t = <fun>
+|}]
+
+let test_lambda a = (fun (module T : Typ) (x : T.t) -> x) (module Int) a
+
+[%%expect{|
+val test_lambda : Int.t -> Int.t = <fun>
+|}]
+
+
+let alpha_equiv (f : (module A : Add) -> A.t -> A.t)
+  : (module T : Add) -> T.t -> T.t = f
+
+[%%expect{|
+val alpha_equiv :
+  ((module A : Add) -> A.t -> A.t) -> (module T : Add) -> T.t -> T.t = <fun>
+|}]
+
+(* Here we test that M is not captured inside the type of f *)
+let apply_weird (module M : Typ) (f : (module M : Typ) -> _) (x : M.t) : M.t =
+  f (module M) x
+
+[%%expect{|
+val apply_weird :
+  (module M : Typ) -> ((module M : Typ) -> M/2.t -> M/2.t) -> M.t -> M.t =
+  <fun>
+|}]
+
+(** From here on we will try applying invalid arguments to functions *)
+
+let f x (module M : Typ) (y : M.t) = (x, y)
+
+[%%expect{|
+val f : 'a -> (module M : Typ) -> M.t -> 'a * M.t = <fun>
+|}]
+
+(* f does not constraint the type of the first argument of the function *)
+let invalid_arg1 = f (module Int)
+
+[%%expect{|
+Line 1, characters 21-33:
+1 | let invalid_arg1 = f (module Int)
+                         ^^^^^^^^^^^^
+Error: The signature for this packaged module couldn't be inferred.
+|}]
+
+(* We gave too many arguments before the module argument, resulting in a type
+error *)
+let invalid_arg2 = f 3 4 (module Int)
+
+[%%expect{|
+Line 1, characters 23-24:
+1 | let invalid_arg2 = f 3 4 (module Int)
+                           ^
+Error: The constant "4" has type "int" but an expression was expected of type
+         "(module Typ)"
+|}]
+
+(* Here we cannot extract the type of m *)
+let invalid_arg3 =
+  let m = (module Int : Typ) in
+  f 3 m 4
+
+[%%expect{|
+Line 3, characters 2-5:
+3 |   f 3 m 4
+      ^^^
+Error: This expression has type "(module M : Typ) -> M.t -> 'a * M.t"
+       but an expression was expected of type "(module Typ) -> 'b"
+       The module "M" would escape its scope
+  Attempted to remove dependency because
+  could not extract path from module argument.
```

Comment:
This error message is unclear, hard to understand, and also the grammar is not quite right.

I think that the error is trying to say something like:

> The function is a module-dependent function types, which must be passed a static module argument "(module M : Typ)". This argument is not static, so the application is invalid.

(or maybe some other wording than "static module"?)

Reply:
I'm happy with this. (I don't know if the the compiler error messages typically use "we", I suspect that we use weird passive forms to avoid making a choice there, but I'm fine with it.) It may be better than what I wrote in other cases -- where the error cause is difference and the explanation I gave becomes wrong.